File: Haskell/Example/MetricExample.hs
{-# LANGUAGE ExistentialQuantification #-}

-- | Example code used to explain the metrics used
-- in this thesis.

import Data.Foldable (traverse_)

class Speaker a where
    speak :: a -> IO ()

newtype Canine = Canine { sound :: String }

instance Speaker Canine where
    speak (Canine sound) = putStrLn sound

type Dog = Canine

type Wolf = Canine

mkDog :: Dog
mkDog = Canine "Woof"

mkWolf :: Wolf
mkWolf = Canine "Awoo"

newtype Cat = Cat { size :: Int }

mkCat :: Int -> Cat
mkCat = Cat

instance Speaker Cat where
    speak (Cat size) = if size > 2
        then rawr
        else meow
      where
        meow = putStrLn "Meow"
        rawr = putStrLn "Rawr"

-- ExistentialQuantification used here
data SomeSpeaker =
    forall a . Speaker a => SomeSpeaker a

instance Speaker SomeSpeaker where
    speak (SomeSpeaker x) = speak x

main :: IO ()
main = do
    let
        zoo =
            [ SomeSpeaker (mkCat 1)
            , SomeSpeaker (mkCat 3)
            , SomeSpeaker mkDog
            , SomeSpeaker mkWolf
            ]
    traverse_ speak zoo

File: Java/Example/MetricExample.java
// Example code used to explain the metrics used
// in this thesis.


interface Speaker {
	public void speak();
}

abstract class SpeaksWithSound implements Speaker {
	protected String sound;

	protected SpeaksWithSound(String sound) {
		this.sound = sound;
	}

	public void speak() {
		System.out.println(this.sound);
	}
}

class Dog extends SpeaksWithSound {
	public Dog() {
		super("Woof");
	}
}

class Wolf extends SpeaksWithSound {
	public Wolf() {
		super("Awoo");
	}
}

class Cat implements Speaker {
	private static final String SMALL_SOUND = "Meow";
	private static final String BIG_SOUND = "Rawr";
	private int size;

	public Cat (int size) {
		this.size = size;
	}

	public void speak() {
		if (size > 2) {
			System.out.println(BIG_SOUND);
		} else {
			System.out.println(SMALL_SOUND);
		}
	}
}


class MetricExample {
	public static void main(String... args) {
		Speaker[] zoo = new Speaker[] {
			new Cat(1),
			new Cat(3),
			new Dog(),
			new Wolf()
		};

		for (Speaker animal : zoo) {
			animal.speak();
		}
	}
}
